c2corg_api/tweens/rate_limiting.py

```diff
@@ -0,0 +1,72 @@
+import logging
+import datetime
+import pytz
+
+from pyramid.httpexceptions import HTTPBadRequest, HTTPTooManyRequests
+from c2corg_api.emails.email_service import get_email_service
+from c2corg_api.views import http_error_handler
+from c2corg_api.models import DBSession
+from c2corg_api.models.user import User
+
+log = logging.getLogger(__name__)
+
+
+def rate_limiting_tween_factory(handler, registry):
+    """ Add a rate limiting protection on write requests.
+    """
+
+    def tween(request):
+
+        log.debug('RATE LIMITING FOR METHOD ' + request.method)
+
+        # Only write requests are considered for rate limiting.
+        if request.method not in ['POST', 'PUT', 'DELETE']:
+            return handler(request)
+
+        if request.authorization is None:
+            # See comment of similar block in jwt_database_validation tween
+            return handler(request)
+
+        user = DBSession.query(User).get(request.authenticated_userid)
+        if user is None:
+            return http_error_handler(
+                HTTPBadRequest('Unknown user'), request)
+
+        now = datetime.datetime.now(pytz.utc)
+        if user.ratelimit_reset is None or user.ratelimit_reset < now:
+            # No window exists or it is expired: create a new one.
+            span = int(registry.settings.get('rate_limiting.window_span'))
+            limit = int(registry.settings.get(
+                'rate_limiting.limit_robot' if user.robot else
+                'rate_limiting.limit_moderator' if user.moderator else
+                'rate_limiting.limit'))
+            user.ratelimit_reset = now + datetime.timedelta(seconds=span)
+            user.ratelimit_limit = limit
```

Comment:
I do not understand why you need to store limit in database as it is calculated from user type and settings ?

Reply:
Right, I have removed this attribute which is indeed not that useful.